pkgs/development/compilers/gcc/common/pre-configure.nix

```diff
@@ -10,3 +10,35 @@ lib.optionalString (hostPlatform.isSunOS && hostPlatform.is64bit) ''
 '' + lib.optionalString (lib.versionOlder version "7" && (langJava || langGo)) ''
   export lib=$out;
 ''
+
+# NOTE 2020/3/18: This environment variable prevents configure scripts from
+# detecting the presence of aligned_alloc on Darwin.  There are many facts that
+# collectively make this fix necessary:
+#  - Nix uses a fixed set of standard library headers on all MacOS systems,
+#    regardless of their actual version.  (Nix uses version 10.12 headers.)
+#  - Nix uses the native standard library binaries for the build system.  That
+#    means the standard library binaries may not exactly match the standard
+#    library headers.
```

Comment:
Which library does this symbol come from? While we do use the native Libsystem we only export a subset of it's symbols to avoid problems like this.

Reply:
It isn't declared in any of the libsystem libraries in the store. It sneaks in from outside the store: `libSystem.B.dylib` is linked to `/usr/lib/system/libsystem_malloc.dylib`, which contains `aligned_alloc`.

```
$ DYLD_PRINT_LIBRARIES=1 ./result/bin/alignedalloc-test 
dyld: loaded: /Users/loncaric/Desktop/alignedalloc/./result/bin/alignedalloc-test
dyld: loaded: /opt/nix/store/b8xv2cjbqlpngwn3fx3hf5pdkaym8f0q-Libsystem-osx-10.12.6/lib/libSystem.B.dylib
dyld: loaded: /opt/nix/store/b8xv2cjbqlpngwn3fx3hf5pdkaym8f0q-Libsystem-osx-10.12.6/lib/system/libsystem_c.dylib
dyld: loaded: /opt/nix/store/b8xv2cjbqlpngwn3fx3hf5pdkaym8f0q-Libsystem-osx-10.12.6/lib/system/libsystem_kernel.dylib
dyld: loaded: /opt/nix/store/b8xv2cjbqlpngwn3fx3hf5pdkaym8f0q-Libsystem-osx-10.12.6/lib/libSystem_internal.dylib
[...]
dyld: loaded: <00190446-C6A8-3EA6-92D1-850EA85C84CE> /usr/lib/system/libsystem_malloc.dylib
[...]
allocated memory at 0x294017a0
```

```
$ fgrep -ir aligned_alloc /usr/lib/system/libsystem_malloc.dylib
Binary file /usr/lib/system/libsystem_malloc.dylib matches
```

I don't have a good mental model of how linkers work, but I'm not terribly surprised that one might discover `aligned_alloc` through this transitive link.

Reply:
Here's the test case I put together, which compiles on Catalina but not on earlier versions of MacOS:

```
$ cat alignedalloc.c 

int main() {
    void* p = aligned_alloc(16, 64);
    printf("allocated memory at %p\n", p);
    return p ? 0 : 1;
}
```
```
$ cat default.nix 

with import <nixpkgs> {};

stdenv.mkDerivation rec {
  name = "alignedalloc-test";
  src = ./.;
  buildPhase = ''
    cc -v alignedalloc.c -o a.out
  '';
  doCheck = true;
  checkPhase = ''
    ./a.out
  '';
  installPhase = ''
    mkdir -p $out/bin
    mv a.out $out/bin/alignedalloc-test
  '';
}
```
```
$ nix-build
[...]
/opt/nix/store/b203gi34fvm1y5v26i4h3vgav9cdqgp8-alignedalloc-test
```

Reply:
@LnL7 Any remaining concerns?

Reply:
I guess, a slightly more robust way to address this impurity would be to come up with a common list of symbols for libsystem_malloc and plug it into the https://github.com/NixOS/nixpkgs/blob/master/pkgs/os-specific/darwin/apple-source-releases/Libsystem/default.nix . A while I made a tool for doing this using travis https://github.com/veprbl/libsystem_symbol_list , it may come useful.